### PR TITLE
Issue/8285 repair schedules blocked resources

### PR DESCRIPTION
**Pull request by Hugo-Inmanta**
# Description

closes https://github.com/inmanta/inmanta-core/issues/8285

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
~~- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )~~
~~- [ ] If this PR fixes a race condition in the test suite, also push the fix to the relevant stable branche(s) (see [test-fixes](https://internal.inmanta.com/development/core/tasks/build-master.html#test-fixes) for more info)~~

### CODE REVIEW

**Review by hlloreda — APPROVED** [review]

File: src/inmanta/deploy/work.py [hunk 1]
```diff
@@ -473,8 +473,10 @@ def _queue_if_ready(self, blocked_deploy: BlockedDeploy) -> None:
 
     def delete_resource(self, resource: ResourceIdStr) -> None:
         """
-        Drop tasks for a given resource when it's deleted from the model. Does not affect dry-run tasks because they
-        do not act on the latest desired state.
+        Drop tasks for a given resource when it was deleted from the model or
+        when we know it can't progress e.g. it is known to be blocked
+        on another resource.
+        Does not affect dry-run tasks because they do not act on the latest desired state.
```
Comment (sanderr) [inline src/inmanta/deploy/work.py:479]: Nice, good clarification!

File: tests/deploy/test_scheduler_agent.py [hunk 2]
```diff
@@ -1679,3 +1679,34 @@ async def test_scheduler_priority(agent: TestAgent, environment, make_resource_m
...
+    agent.executor_manager.reset_executor_counters()
+
+    # mark r1 as blocked and trigger a repair
+    agent.scheduler._state.resource_state[rid1].blocked = BlockedStatus.YES
+
+    await agent.scheduler.repair(reason="Test triggered global repair")
+    await retry_limited(utils.is_agent_done, timeout=5, scheduler=agent.scheduler, agent_name="agent1")
+
+    # Only r2 was deployed
+    assert agent.executor_manager.executors["agent1"].execute_count == 1
```
Comment (sanderr) [inline tests/deploy/test_scheduler_agent.py:1712]: The comment says "only r2" but you only assert r1.

File: tests/deploy/test_scheduler_agent.py [hunk 3]
```diff
@@ -1679,3 +1679,34 @@ async def test_scheduler_priority(agent: TestAgent, environment, make_resource_m
...
+    rid1 = ResourceIdStr("test::Resource[agent1,name=1]")
+    rid2 = ResourceIdStr("test::Resource[agent1,name=2]")
+
+    resources = {
+        rid1: make_resource_minimal(rid=rid1, values={"value": "r1_value"}, requires=[]),
+        rid2: make_resource_minimal(rid=rid2, values={"value": "r2_value"}, requires=[]),
+    }
+
+    await agent.scheduler._new_version(version=1, resources=resources, requires=make_requires(resources))
+    await retry_limited(utils.is_agent_done, timeout=5, scheduler=agent.scheduler, agent_name="agent1")
```
[the remark this answers is not in the document]
Comment (Hugo-Inmanta) [reply tests/deploy/test_scheduler_agent.py:1698]: 20 times faster :rocket:

File: tests/deploy/test_scheduler_agent.py (hunk 2 above)
Comment (Hugo-Inmanta) [reply tests/deploy/test_scheduler_agent.py:1712]: Unless I'm missing something, r1 is blocked and only r2 is being redeployed.
Maybe I can improve the wording ? Or change the test to explicitly check that r2 is being deployed (I don't know if there's an easy way to check for this)?

Comment (sanderr) [reply tests/deploy/test_scheduler_agent.py:1714]: My comment was not on point. I had missed that you had both resources on the same agent, so indeed your assertion is correct.

File: tests/deploy/test_scheduler_agent.py [hunk 4]
```diff
@@ -1755,7 +1757,7 @@ async def test_repair_does_not_trigger_for_blocked_resources(agent: TestAgent, m
     }
 
     await agent.scheduler._new_version(version=1, resources=resources, requires=make_requires(resources))
-    await retry_limited(utils.is_agent_done, timeout=5, scheduler=agent.scheduler, agent_name="agent1")
+    await retry_limited_fast(utils.is_agent_done, timeout=5, scheduler=agent.scheduler, agent_name="agent1")
```
Comment (sanderr) [inline tests/deploy/test_scheduler_agent.py:1760]:
```suggestion
    await retry_limited_fast(utils.is_agent_done, scheduler=agent.scheduler, agent_name="agent1")
```

File: tests/deploy/test_scheduler_agent.py [hunk 5]
```diff
@@ -1766,7 +1768,7 @@ async def test_repair_does_not_trigger_for_blocked_resources(agent: TestAgent, m
     agent.scheduler._state.resource_state[rid1].blocked = BlockedStatus.YES
 
     await agent.scheduler.repair(reason="Test triggered global repair")
-    await retry_limited(utils.is_agent_done, timeout=5, scheduler=agent.scheduler, agent_name="agent1")
+    await retry_limited_fast(fun=utils.is_agent_done, timeout=5, scheduler=agent.scheduler, agent_name="agent1")
```
Comment (sanderr) [inline tests/deploy/test_scheduler_agent.py:1771]:
```suggestion
    await retry_limited_fast(fun=utils.is_agent_done, scheduler=agent.scheduler, agent_name="agent1")
```

**Review by sanderr — APPROVED** [review]

**Pull request closed without merging**